Added endpoints for different transaction types.

WarFire5 opened the pull request.

WarFire5 committed:
Added AddTransactionValidator

WarFire5 committed:
Added validators

WarFire5 committed:
Added add-transaction controllers

WarFire5 committed:
Removed Authorization

WarFire5 committed:
Added balance and accounts checks

WarFire5 committed:
Merge remote-tracking branch 'origin/addfilter_addbalancecontroller' into Creating_controllers_for_adding_transactions

WarFire5 committed:
Post-merge adjustment

WarFire5 committed:
The order of methods has been clarified.

WarFire5 committed:
Added calculation of transfer-deposit amount by currency rate.

WarFire5 committed:
Added check for account currency type.

Negodyaj commented on TransactionStore.API/Controllers/TransactionsController.cs:
```diff
@@ -8,9 +11,54 @@ namespace TransactionStore.API.Controllers;
+
+    // получаем тип валюты аккаунта по его Id
+    [HttpGet("/currency/{id}")]
+    public ActionResult<CurrencyTypeResponse> GetCurrencyTypeByAccountId(Guid id)
```
not needed

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+        _addTransferValidator = addTransferValidator;
+    }
+
+    public CurrencyTypeResponse GetCurrencyTypeByAccountId(Guid id)
```
not needed

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+
+        if (validationResult.IsValid)
+        {
+            var сurrencyTypeResponse = GetCurrencyTypeByAccountId(request.AccountId);
```
not needed

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+                return _transactionsRepository.AddDepositTransaction(transaction);
+            }
+
+            throw new InappropriateCurrencyTypeException("Тип валюты не совпадает с типом валюты аккаунта.");
```
not needed

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+                    return _transactionsRepository.AddWithdrawTransaction(transaction);
+                }
+
+                throw new NotEnoughMoneyException("На балансе недостаточно средств для проведения операции.");
```
not needed

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+                throw new NotEnoughMoneyException("На балансе недостаточно средств для проведения операции.");
+            }
+
+            throw new InappropriateCurrencyTypeException("Тип валюты не совпадает с типом валюты аккаунта.");
```
not needed

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+        return _mapper.Map<AccountBalanceResponse>(accountBalance);
+    }
+
+    public Guid AddDepositTransaction(DepositWithdrawRequest request)
```
process both Deposit and Withdraw transactions in 1 method

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +18,191 @@ public class TransactionsService : ITransactionsService
+
+        if (validationResult.IsValid)
+        {
+            if (request.AccountFromId != request.AccountToId && request.CurrencyFromType != request.CurrencyToType)
```
not needed

Negodyaj commented on TransactionStore.Core/Data/DictionaryOfCoefficients.cs:
```diff
@@ -1,82 +1,84 @@
+            { "USDCNY", 7.22m },
+            { "USDRSD", 107.50m },
+            { "USDBGN", 1.80m },
+            { "USDARS", 883.67m },
```
these are needed:
{ "USDRUB", 90.89m },
            { "USDEUR", 0.92m },
            { "USDJPY", 155.66m },
            { "USDCNY", 7.22m },
            { "USDRSD", 107.50m },
            { "USDBGN", 1.80m },
            { "USDARS", 883.67m },

other to be removed

Negodyaj commented on TransactionStore.Core/Data/DictionaryOfCoefficients.cs:
```diff
@@ -1,82 +1,84 @@
-namespace TransactionStore.Core.Data
+namespace TransactionStore.Core.Data;
+
+public class DictionaryOfCoefficients
```
CurrencyRatesProvider

Negodyaj commented on TransactionStore.Core/Data/DictionaryOfCoefficients.cs:
```diff
@@ -1,82 +1,84 @@
-        {
-            return _rate[currency1 + currency2];
-        }
+    public decimal GetRate(string currency1, string currency2)
```
GetRate("CNY", "JPY");

Negodyaj commented on TransactionStore.Core/Models/Transactions/Requests/DepositWithdrawRequest.cs:
```diff
@@ -1,12 +1,10 @@
 namespace TransactionStore.Core.Models.Transactions.Requests;
 
-public class AddTransactionRequest
+public class DepositWithdrawRequest
```
remove Date from request

Negodyaj commented on TransactionStore.Core/Models/Transactions/Requests/TransferRequest.cs:
```diff
@@ -0,0 +1,13 @@
+    public CurrencyType CurrencyFromType { get; set; }
+    public CurrencyType CurrencyToType { get; set; }
+    public int Amount { get; set; }
+    public DateTime Date { get; set; }
```
remove date from request

Negodyaj commented on TransactionStore.Core/Validators/AddDepositWithdrawValidator.cs:
```diff
@@ -0,0 +1,41 @@
+            .Must(guid => Regex.IsMatch(guid.ToString(), @"^[{]?[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}[}]?$"))
+            .WithMessage("Некорректный формат GUID. GUID должен быть в формате XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX, " +
+                         "где X — это шестнадцатеричный символ. Допустимы фигурные скобки в начале и конце строки.");
+
```
.NotEmpty().WithMessage("GUID не должен быть значением по умолчанию (default).")

Negodyaj commented on TransactionStore.Core/Validators/AddDepositWithdrawValidator.cs:
```diff
@@ -0,0 +1,41 @@
+
+        RuleFor(t => t.Amount)
+            .NotEmpty().WithMessage("Поле не может быть пустым. Введите сумму операции.")
+            .NotNull().WithMessage("Поле не может быть null. Введите сумму операции.")
```
not needed

Negodyaj commented on TransactionStore.Core/Validators/AddDepositWithdrawValidator.cs:
```diff
@@ -0,0 +1,41 @@
+            .GreaterThanOrEqualTo(1).WithMessage("Сумма не должна быть меньше 1.")
+            .LessThanOrEqualTo(1000000).WithMessage("Превышен максимальный лимит на операцию.");
+
+        RuleFor(t => t.Date)
```
not needed

Negodyaj commented on TransactionStore.DataLayer/Migrations/TransactionStoreContextModelSnapshot.cs:
```diff
@@ -37,8 +37,8 @@ protected override void BuildModel(ModelBuilder modelBuilder)
-                    b.Property<int>("Amount")
-                        .HasColumnType("integer")
+                    b.Property<decimal>("Amount")
+                        .HasColumnType("numeric")
```
NUMERIC(precision, scale)

Negodyaj commented on TransactionStore.DataLayer/Migrations/20240522162435_CreatedDataBase.cs:
```diff
@@ -25,7 +25,7 @@ protected override void Up(MigrationBuilder migrationBuilder)
                     transaction_type = table.Column<TransactionType>(type: "transaction_type", nullable: false),
                     currency_type = table.Column<CurrencyType>(type: "currency_type", nullable: false),
-                    amount = table.Column<int>(type: "integer", nullable: false),
+                    amount = table.Column<decimal>(type: "numeric", nullable: false),
```
NUMERIC(precision, scale)

Negodyaj commented on TransactionStore.DataLayer/Repositories/TransactionsRepository.cs:
```diff
@@ -9,4 +10,39 @@ public class TransactionsRepository : BaseRepository, ITransactionsRepository
     {
     }
+
+    public TransactionDto GetCurrencyTypeByAccountId(Guid id)
```
not needed

Negodyaj commented on TransactionStore.DataLayer/Repositories/TransactionsRepository.cs:
```diff
@@ -9,4 +10,39 @@ public class TransactionsRepository : BaseRepository, ITransactionsRepository
+        return transaction.Id;
+    }
+
+    public Guid AddWithdrawTransaction(TransactionDto transaction)
```
to be deleted

WarFire5 committed:
Purged

WarFire5 committed:
Updated database

WarFire5 committed:
Merged deposit and withdrawal into one method.

WarFire5 committed:
Added logic with currency rates for transfer transactions.

WarFire5 committed:
AddDepositWithdrawTransaction and AddTransferTransaction methods were simplified.

WarFire5 committed:
Added TransactionsControllerTests

WarFire5 committed:
Added TransactionsRepositoryTests

WarFire5 committed:
Fixed AddDepositWithdrawTransaction method

Negodyaj commented on TransactionStore.API.Tests/TransactionsControllerTests.cs:
```diff
@@ -0,0 +1,134 @@
+{
+    public class TransactionsControllerTests
+    {
+        private TransactionsController CreateController(Mock<ITransactionsService> mockService)
```
not needed

Negodyaj commented on TransactionStore.API/Controllers/TransactionsController.cs:
```diff
@@ -8,9 +12,64 @@ namespace TransactionStore.API.Controllers;
+    [HttpPost("/deposit")]
+    public ActionResult<Guid> AddDepositTransaction([FromBody] DepositWithdrawRequest request)
+    {
+        if (request == null || request.AccountId == Guid.Empty)
```
передай пустой боди и посмотри)
а аккаунт проверяется в валидаторе

Negodyaj commented on TransactionStore.API/Extensions/ConfigureDataBase.cs:
```diff
@@ -1,12 +1,13 @@
 
 public static class DataBaseExtensions
 {
+    [Obsolete]
```
не надо

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +16,107 @@ public class TransactionsService : ITransactionsService
+
+        if (validationResult.IsValid)
+        {
+            TransactionDto transaction;
```
= _mapper.Map<TransactionDto>(request);

Negodyaj commented on TransactionStore.Business/Services/TransactionsService.cs:
```diff
@@ -11,14 +16,107 @@ public class TransactionsService : ITransactionsService
+    private TransactionDto CreateDepositTransaction(TransferRequest request)
+    {
+        var dictionaryOfCoefficients = new CurrencyRatesProvider();
+        var rateToUSD = dictionaryOfCoefficients.GetRateToUsd(request.CurrencyFromType.ToString());
```
перенеси ToString() внутрь этого метода

WarFire5 committed:
Merge remote-tracking branch 'origin/master' into Creating_endpoints_for_adding_transactions

WarFire5 committed:
Fixed after merge

WarFire5 committed:
Updated GetTransactionsByAccountId_GuidSent_ListTransactionsDtoReceived

Updated without using virtual GetTransactionsByAccountId_GuidSent_ListTransactionsDtoReceived

WarFire5 committed:
Updated TransactionsControllerTests

WarFire5 committed:
Purged TransactionsControllerTests

WarFire5 committed:
Modified in accordance with comments

WarFire5 committed:
Added TransactionsServiceTests, updated TransactionsRepositoryTests

Added TransactionsServiceTests, updated TransactionsRepositoryTests without using Moq
